### PR TITLE
Add bm25 service

### PR DESCRIPTION
**Pull request by bhavani105**
# Pull Request      

- Implement BM25 (Pyserini) Retriever and Reader component
- Add support for BM25 indexing and retrieval in the services layer

```Closes #(issue)```       

Notes:
- Replace `(issue)` **above** ↑↑↑ with the issue this PR closes to automatically link the two.
This must be done when the PR is created.
- Add multiple `Closes #(issue)` as needed.
- If this PR is work towards but does not close an issue, simply tag the issue without mentioning `Closes`.

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

```(description)```     

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

File: primeqa/pipelines/components/base.py [hunk 1]
```diff
@@ -89,5 +102,18 @@ def __hash__(self) -> int:
         raise NotImplementedError
 
     @abstractmethod
-    def retrieve(self, input_texts: List[str], *args, **kwargs):
+    def retrieve(self, queries: List[str], *args, **kwargs):
```
Comment (kpfadnis) [inline primeqa/pipelines/components/base.py:105]: If we are changing signature of `retrieve` method, we need to follow that signature change in both `sparse` and `dense` retrievers.

File: primeqa/pipelines/components/retriever/sparse.py [hunk 2]
```diff
@@ -29,22 +31,55 @@ class BM25Retriever(RetrieverComponent):
...
+            f"{self.__class__.__name__}::{json.dumps({k: v for k, v in vars(self).items() if k in hashable_fields}, sort_keys=True)}"
+        )
 
     def load(self, *args, **kwargs):
-        pass
+        self._searcher = PyseriniRetriever(self._index_path)
 
     def retrieve(self, input_texts: List[str], *args, **kwargs):
-        pass
+        qids = [str(idx) for  idx, query in enumerate(input_texts) ]
```
Comment (kpfadnis) [inline primeqa/pipelines/components/retriever/sparse.py:77]: Shouldn't this change to `queries` as per change in the `base.py` for `retriever` component?

**Review by kpfadnis — CHANGES_REQUESTED** [review]
@bhavani105 Please verify that `retrieve` signature change is fixed across all retrievers and necessary notebooks as well. Thanks

**Review by kpfadnis — APPROVED** [review]
LGTM. Nice work!

File: primeqa/pipelines/components/indexer/dense.py [hunk 3]
```diff
@@ -139,3 +139,6 @@ def index(self, collection: Union[List[dict], str], *args, **kwargs):
             collection,
             overwrite="overwrite" in kwargs and kwargs["overwrite"],
         )
+
```
Comment (franzmpub) [inline primeqa/pipelines/components/indexer/dense.py:142]: How about other dense IR engines?

File: primeqa/pipelines/components/indexer/sparse.py [hunk 4]
```diff
@@ -17,9 +22,49 @@ class BM25Indexer(IndexerComponent):
...
+            )
+                
+        self._indexer.index_collection(collection = collection, index_path=self._index_path, 
+            fieldnames=None, 
+            overwrite="overwrite" in kwargs and kwargs["overwrite"],
+            threads=kwargs["num_workers"] if "num_workers" in kwargs else 1, 
+            additional_index_cmd_args=kwargs["additional_index_args"] if "additional_index_args" in kwargs 
+                else '--storePositions --storeDocvectors --storeRaw' )
+    
+    def get_engine_type(self) -> str:
```
Comment (franzmpub) [inline primeqa/pipelines/components/indexer/sparse.py:65]: Similarly as with dense IR (but more hypothetically): how about other sparse IR engines?

**Review by franzmpub — APPROVED** [review]
LGTM

**Review by hrojek — APPROVED** [review]
Tested end to end, with primeqa-ui and primeqa-orchestrator.

**Pull request merged**